src/http/one/TeChunkedParser.cc

```diff
@@ -115,55 +116,61 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
+        if (tok.skipAll(NonLF) && tok.skip('\n'))
+            throw TexcHere("cannot parse chunk extension"); // <garbage> CR*LF
+
+        return false; // need more data to finish parsing extension
+    }
+    return false; // need data to start parsing extension or CRLF
+}
+
+bool
+Http::One::TeChunkedParser::parseChunkExtension(::Parser::Tokenizer &tok, bool skipKnown)
 {
     SBuf ext;
     SBuf value;
-    while (
-        ParseBws(tok) && // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
-        tok.skip(';') &&
-        ParseBws(tok) && // Bug 4492: ICAP servers send SP before chunk-ext-name
-        tok.prefix(ext, CharacterSet::TCHAR)) { // chunk-ext-name
-
-        // whole value part is optional. if no '=' expect next chunk-ext
-        if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
-
-            if (!skipKnown) {
-                if (ext.cmp("use-original-body",17) == 0 && tok.int64(useOriginBody, 10)) {
-                    debugs(94, 3, "Found chunk extension " << ext << "=" << useOriginBody);
-                    buf_ = tok.remaining(); // parse checkpoint
-                    continue;
-                }
-            }
-
-            debugs(94, 5, "skipping unknown chunk extension " << ext);
+    if (!(ParseBws(tok) && // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
+            tok.skip(';') &&
+            ParseBws(tok) && // Bug 4492: ICAP servers send SP before chunk-ext-name
+            tok.prefix(ext, CharacterSet::TCHAR))) // chunk-ext-name
+        return false;
 
-            // unknown might have a value token or quoted-string
-            if (tok.quotedStringOrToken(value) && !tok.atEnd()) {
-                buf_ = tok.remaining(); // parse checkpoint
-                continue;
+    // whole value part is optional. if no '=' expect next chunk-ext
+    if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
+        if (!skipKnown) {
+            if (ext.cmp("use-original-body",17) == 0 && tok.int64(useOriginBody, 10)) {
+                debugs(94, 3, "Found chunk extension " << ext << "=" << useOriginBody);
+                return true;
             }
-
-            // otherwise need more data OR corrupt syntax
-            break;
         }
 
-        if (!tok.atEnd())
-            buf_ = tok.remaining(); // parse checkpoint (unless there might be more token name)
-    }
+        debugs(94, 5, "skipping unknown chunk extension " << ext);
 
-    if (skipLineTerminator(tok)) {
-        buf_ = tok.remaining(); // checkpoint
-        // non-0 chunk means data, 0-size means optional Trailer follows
-        parsingStage_ = theChunkSize ? Http1::HTTP_PARSE_CHUNK : Http1::HTTP_PARSE_MIME;
-        return true;
-    }
+        // unknown might have a value token or quoted-string
+        if (quotedStringOrToken(tok, value) && !tok.atEnd())
+            return true;
 
-    return false;
+        return false;
+    }
+    return true;
```

Comment:
This `true` return value or the condition preceding it does not look right. For example, If the remaining input contains nothing or just BWS, then the return value should be `false`, not `true` -- we have not parsed the current extension yet, and the rest of that current extension (i.e., extension value) may still be coming.

In other words, the extension "=value" part is indeed optional, but we still need to wait for it if it might get it during the next I/Os.

Reply:
This concern remains unaddressed IMO, and the new `parsed the valueless chunk-ext` comment may be lying: For example, if our input contains `;name=`, then the method should return false, not true -- we have not parsed the current extension yet, and the rest of that current extension (i.e., BWS and/or the extension value) may still be coming.

Reply:
> This concern remains unaddressed IMO, and the new parsed the valueless chunk-ext comment may be lying: For example, if our input contains `;name=`

Disagree, the new code should work properly. Your particular `;name=` example: the execution will pass into the `if (tok.skip('=') && ParseBws(tok)) { }` block and returns false in the end.
Are there any other examples which could be misinterpreted by the new code?

Reply:
> Your particular ;name= example: the execution will pass into the if (tok.skip('=') && ParseBws(tok)) { } block and returns false in the end.

I see. ParseBws() is also broken -- it returns true when there are no characters in the input buffer. That is incorrect because it allows the code to mistakenly think that BWS was handled already (and switch to the next parsing step/phase) when, in fact, BWS will appear in the input stream after the next I/O. Please fix both bugs.

Reply:
Done at 781dbe5.

Reply:
> BWS will appear in the input stream after the next I/O

However, this specific case should be handled properly, I think: each chunk-ext parsing step starts with ParseBws(), thus skipping unexpected whitespaces received at each I/O.